### ISSUE TITLE
`enum_variant_names` does not consider encapsulated type

### ISSUE THREAD

**Issue by chrysn** [rust-lang/rust-clippy#13637]
Clippy is unhappy with this perfectly reasonable enum:

```rust
pub enum Error {
    RecvError(udp::RecvError),
    SendError(udp::SendError),
    BindError(udp::BindError),
    AddressFamilyUnavailable,
}
```

The lint `enum_variant_names` fires because `RecvError`, `SendError` and `BindError` variants are named like the type. That's the fundamental point of the lint, but it fails to consider information that'd be available to the linter: That that substring is also part of the encapsulated type.

I'm not well enough familiar with clippy to come up with code, but I'd like to suggest the following addition to the lint's rules:

> If the variant contains a single item, and any component of the contained item's path is identical to the variant's name, then the lint does not trigger.

By including the full path, it would even protect a hypothetical `enum Cbor { MiniCbor(minicbor::Something), SerdeCbor(serde_cbor::Something) }`; I can't tell how much going-out-of-the-way it would need to catch `SerdeCbor` to be justified by `serde_cbor`, so maybe a simpler version that just doesn't trigger on the real example above suffices.

**Comment by hikari-no-yume** [comment]
Despite this being merged, we just hit this in [the CI for touchHLE](https://github.com/touchHLE/touchHLE/actions/runs/19274342582/job/55110390148#step:21:210):

<img width="621" height="462" alt="Image" src="https://github.com/user-attachments/assets/afd8ff18-7046-40a4-bfa0-79d63f1bafe6" />

Has this commit not reached stable yet or something?

**Comment by hikari-no-yume** [comment]
It's supposedly on the rust-1.91.0, rust-1.90.0 and rust-1.89.0 branches, and given the help link in that screenshot, presumably it's the rust-1.19.0 compiler that's complaining… am I missing something?

**Comment by aaron-ang** [comment]
@hikari-no-yume, your error is not related to this issue or the PR. Your enum variants have the same prefix, so that warning triggers first before the inner tuple matching.

The issue description also states, _"The lint enum_variant_names fires because RecvError, SendError and BindError variants are named like the type. That's the fundamental point of the lint..."_

**Comment by hikari-no-yume** [comment]
So what you're saying is that the variant names matching the inner type names only overrides the lint for the name of the enum being part of the name of the variant, and not the lint for the names of the variants having a shared prefix? I see. I didn't realise these were two different lints! From my perspective, the override would make sense for the latter situation too…

**Comment by aaron-ang** [comment]
> So what you're saying is that the variant names matching the inner type names only overrides the lint for the name of the enum being part of the name of the variant, and not the lint for the names of the variants having a shared prefix? I see. I didn't realise these were two different lints! From my perspective, the override would make sense for the latter situation too…

It is part of the same lint, but the lint has multiple sequential checks. The lint first checks that the variant names don't share the same prefix/suffix (which fails in your example). If that check passes, it will then exclude variants with identical inner path components.

**Comment by aaron-ang** [comment]
I think we compare among the variant names, not with the enum name.

**Comment by hikari-no-yume** [comment]
I see. It's an unfortunate heuristic because it makes two assumptions that are wrong in this case: 1) that the shared prefix between members must have something to do with the enclosing enum, for example an abbreviation (it is not), 2) that the shared prefix means we are importing the enum variants directly rather than writing out the enum name (we are not).

**Comment by aaron-ang** [comment]
@hikari-no-yume I may have oversimplified the lint. I looked at the code again and can confirm that it compares against the enum name first, then amongst the enum variants. Regardless, I agree that the assumptions made do not fit your use case. Would removing the `CG` prefix affect readability? `CG` represents core graphics in your project. It seems to me that you could remove the prefix to make it more generic.

**Comment by hikari-no-yume** [comment]
The CG prefix comes from an Apple system API that's native to C/Objective-C/Swift. There's a huge number of Apple types with this kind of naming; NS, UI and CF are among the most common prefixes. As it happens, we're reimplementing those in our project, so we technically could use whatever name we want. But the point is that the name of the type is notionally not something we have control over, the enum variant is just named after the type, and removing the prefix wouldn't make sense because there is no connection between the enum's name and the prefix on the variants.
